### PR TITLE
[#277] tracer verbosity in ToObject instances

### PR DESCRIPTION
**Pull request by CodiePP**
description
-----------

- [x] control verbosity of data structures via `TracerVerbosity` and instances of `ToObject`

checklist
---------

- [x] compiles (`cabal new-clean; cabal new-build`)
- [x] tests run successfully (`cabal new-test`)
- [x] documentation added and created (`cd docs; nix-shell --run make`)
- [ ] link to an issue
- [x] link to an epic
- [x] add estimate points
- [x] add milestone (the same as the linked issue)

### CODE REVIEW

File: iohk-monitoring/test/Cardano/BM/Test/Structured.lhs [hunk 1]
```diff
@@ -96,4 +101,21 @@ logStructured = do
         ("verify traced structure with privacy annotation Confidential: " ++ (show ms))
         (Confidential == privacy (loMeta (ms !! 0)))
 
+logStructuredStdout :: Assertion
```
Comment (CodiePP) [inline iohk-monitoring/test/Cardano/BM/Test/Structured.lhs]: output of this test looks like this:
```
[iohk.logStructured:Notice:3] [2019-08-08 21:29:50.91 UTC] integer = 42
[iohk.logStructured:Debug:3] [2019-08-08 21:29:50.91 UTC] {"kind":"Pet","age":8,"name":"bella"}
[iohk.logStructured:Debug:3] [2019-08-08 21:29:50.91 UTC] {"kind":"Pet","age":8,"name":"bella"}
[iohk.logStructured:Debug:3] [2019-08-08 21:29:50.91 UTC] {"kind":"Pet"}
[iohk.logStructured:Notice:3] [2019-08-08 21:29:50.91 UTC] 42
[iohk.logStructured:Debug:3] [2019-08-08 21:29:50.91 UTC] {"kind":"Pet"}
```

File: iohk-monitoring/src/Cardano/BM/Data/LogItem.lhs [hunk 2]
```diff
@@ -111,10 +111,17 @@ instance FromJSON LOMeta where
...
-           <*> (pack . show <$> (liftIO myThreadId))
+    LOMeta <$> liftIO getCurrentTime
+           <*> (cleantid <$> liftIO myThreadId)
            <*> pure sev
            <*> pure priv
+  where
+    cleantid threadid =
+        let stid = pack $ show threadid
+        in
+        if (isPrefixOf "ThreadId " stid)
```
Comment (njd42) [inline iohk-monitoring/src/Cardano/BM/Data/LogItem.lhs]: I understand the issue here - we want to strip off extraneous text - I've had a look at `base` package and I can't see any alternative interface.

However, the use of a '9' as a constant here worries me (I don't like constants in code other that '0' or '1'!).

Could I suggest using something like `stripPrefix` using a named constant e.g.
  `threadIdText = "ThreadId "
    condStripPrefix p s = maybe s id . stripPrefix p s
    clentId  = condStripPrefix threadIdText threadid . pack . show`

hope that makes sense - gets away from derived constants, makes the constant dependency clear

File: iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs [hunk 3]
```diff
@@ -150,19 +179,23 @@ instance ToObject Foo
...
+class ToObject a where
+    toObject :: TracingVerbosity -> a -> Object
+    default toObject :: ToJSON a => TracingVerbosity -> a -> Object
+    toObject _ v = case toJSON v of
         Object o     -> o
         s@(String _) -> HM.singleton "string" s
         _            -> mempty
 
+mkObject :: ToObject a => [(Text, a)] -> HM.HashMap Text a
+mkObject = HM.fromList
```
Comment (denisshevchenko) [inline iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs]: Do we really need an alias for `HM.fromList`? :-)

Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs]: it's not an alias as the types here are more specific.

**Pull request merged**